gemrb/core/Game.cpp

```diff
@@ -1734,43 +1753,78 @@ bool Game::RestParty(int checks, int dream, int hp)
 // heal on rest and similar
 void Game::CastOnRest()
 {
+	typedef std::multimap<ieWord,HealingResource> RestSpells;
+	typedef std::vector<Injured> RestTargets;
+
 	ieDword tmp = 0;
 	core->GetDictionary()->Lookup("Heal Party on Rest", tmp);
 	int specialCount = core->GetSpecialSpellsCount();
-
+	RestTargets wholeparty;
 	if (tmp && specialCount != -1) {
 		int ps = GetPartySize(true);
 		int ps2 = ps;
-		std::vector<Actor *> injurees;
 		for (int idx = 1; idx <= ps; idx++) {
 			Actor *tar = FindPC(idx);
-			if (tar && tar->GetStat(IE_HITPOINTS) < tar->GetStat(IE_MAXHITPOINTS)) {
-				injurees.push_back(tar);
+			ieWord hpneeded=tar->GetStat(IE_MAXHITPOINTS) - tar->GetStat(IE_HITPOINTS);
+			if (tar && hpneeded > 0) {
+				wholeparty.push_back(Injured(hpneeded,tar));
+			} else {
+				wholeparty.push_back(Injured(0,tar));
 			}
-		}
-		if (injurees.empty()) injurees = PCs; // enable some nonhealing magic too
-
+			
+		}
+		// Following algorithm works thus:
+		// - If at any point there are no more injured party members, stop
+		// (amount of healing done is an estimation)
+		// - cast party members' all heal-all spells
+		// - repeat:
+		//       cast the most potent healing spell on the most injured member
 		SpecialSpellType *special_spells = core->GetSpecialSpells();
+		std::sort(wholeparty.begin(),wholeparty.end());
+		specialCount = core->GetSpecialSpellsCount();
+		RestSpells healingspells;
 		while (specialCount--) {
-			if (special_spells[specialCount].flags & SP_REST) {
-				if (injurees.empty()) injurees = PCs; // enable some nonhealing magic too
-				// check each party member for the spell
+			// Cast multi-target healing spells
+			if (special_spells[specialCount].flags & SP_HEAL_ALL) {
```

Comment:
please include SP_REST too, as this bit may get other meanings in other combinations sometime.